# Introduction to ML — Exercise 1: Descriptive Statistics

**Course:** Introduction to Machine Learning and Applied Artificial Intelligence
**Institution:** High School of Digital Culture, ITMO University



<a href="https://colab.research.google.com/github/tutsilianna/Introduction_to_ML_and_Advanced_ML_Python/blob/main/Introduction/Introduction_%7C_Expercise_1.ipynb" target="_blank">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

This notebook is a warm-up exercise before we get into actual machine learning models.
The goal is to practice the basic descriptive statistics that show up everywhere in
data analysis and ML: the sample **mean**, **median**, **variance**, and **standard
deviation**. We also practice a very common first step in any data project — cleaning
a dataset by removing rows we don't want to include in the analysis.

## Task

We are given a file with the average monthly salary and population of Russian regions
as of January 1, 2019 (source: Russian Federal State Statistics Service). The original
file can be downloaded here:
[salary_and_population.csv](https://drive.google.com/file/d/1YrEIn6JNWW0RG0r8nV0P6S55TBXcxSna/view?usp=drive_link).

Two regions — **Vologda Region** and **Murmansk Region** — were included in the file by
mistake and must be excluded before we compute any statistics.

Using the cleaned sample, we need to find:

1. The sample **mean** of the salary
2. The sample **median** of the salary
3. The **variance** estimate $S^2$ of the salary
4. The **standard deviation** $\sigma$ of the salary


## 0. Setup

Only two libraries are needed for this exercise:

- **numpy** — for numerical helper functions (not strictly required here since pandas
  already exposes `.mean()`, `.median()`, etc., but it's good practice to have it
  available for any array-level math).
- **pandas** — for loading and manipulating the tabular data.


In [1]:
import numpy as np
import pandas as pd


## 1. Load the data

> **Note on data access:** the original exercise ran on Google Colab and loaded the
> file from the author's personal Google Drive with `drive.mount(...)`. That approach
> only works inside Colab and isn't reproducible on GitHub, so this version instead
> expects the CSV to be available locally at `data/salary_and_population.csv`.
>
> To reproduce the notebook:
> 1. Download the file from the
>    [link above](https://drive.google.com/file/d/1YrEIn6JNWW0RG0r8nV0P6S55TBXcxSna/view?usp=drive_link).
> 2. Place it in a `data/` folder next to this notebook (or update `DATA_PATH` below).


In [9]:
DATA_PATH = "data/salary_and_population.csv"

data = pd.read_csv(DATA_PATH, encoding="utf-8", delimiter=",")

# The source file contains one column with no header (an artifact of how it was
# exported), which pandas reads in as "Unnamed: 4". It carries no information, so
# we drop it right away.
data.drop(columns=["Unnamed: 4"], inplace=True)

data.head()


,Region_RU,Region_EN,AVG_Salary,Population
0,Алтайский край,Altai Region,25960,218866
1,Амурская область,Amur Region,43156,793194
2,Архангельская область,Arkhangelsk Region,47526,1144119
3,Астраханская область,Astrakhan Region,32166,1014065
4,Белгородская область,Belgorod Region,31163,1547418


## 2. Inspect the region names

Before filtering, let's check the full list of unique region names in the (Russian)
`Region_RU` column so we know exactly how "Vologda Region" and "Murmansk Region" are
spelled in the source data.


In [10]:
data["Region_RU"].unique()


array(['Алтайский край', 'Амурская область', 'Архангельская область',
       'Астраханская область', 'Белгородская область', 'Брянская область',
       'Владимирская область', 'Волгоградская область',
       'Вологодская область', 'Воронежская область', 'г Москва',
       'г Санкт-Петербург', 'г Севастополь', 'Еврейская АО',
       'Забайкальский край', 'Ивановская область', 'Иркутская область',
       'Кабардино-Балкарская Республика', 'Калининградская область',
       'Калужская область', 'Камчатский край',
       'Карачаево-Черкесская Республика', 'Кемеровская область',
       'Кировская область', 'Костромская область', 'Краснодарский край',
       'Красноярский край', 'Курганская область', 'Курская область',
       'Ленинградская область', 'Липецкая область', 'Магаданская обл',
       'Московская обл', 'Мурманская обл', 'Ненецкий АО',
       'Нижегородская область', 'Новгородская область',
       'Новосибирская область', 'Омская область', 'Оренбургская область',
       'Орловская о

## 3. Remove the two excluded regions

The task states that Vologda Region (`Вологодская область`) and Murmansk Region
(`Мурманская обл`) were mistakenly included and must be dropped before we compute any
statistics.


In [11]:
data = data.loc[
    (data["Region_RU"] != "Вологодская область")
    & (data["Region_RU"] != "Мурманская обл")
]

# Sanity check: the sample size should have shrunk by exactly two rows.
data.shape


(83, 4)

## 4. Compute the descriptive statistics

With the cleaned sample in hand, we can compute the four requested statistics directly
with pandas' built-in aggregation methods. Note that `.var()` and `.std()` use
`ddof=0` (the *population* formula, dividing by $n$) rather than the default `ddof=1`
(the *sample* formula, dividing by $n-1$), matching how the exercise defines the
variance estimate $S^2$.


In [12]:
mean_salary = round(data["AVG_Salary"].mean(), 2)
print(f"Sample mean of the salary: {mean_salary}")


Sample mean of the salary: 38658.04


In [13]:
median_salary = round(data["AVG_Salary"].median(), 2)
print(f"Sample median of the salary: {median_salary}")


Sample median of the salary: 32473.0


In [14]:
variance_salary = round(data["AVG_Salary"].var(ddof=0), 2)
print(f"Variance estimate of the salary (S^2): {variance_salary}")


Variance estimate of the salary (S^2): 270512026.47


In [15]:
std_salary = round(data["AVG_Salary"].std(ddof=0), 2)
print(f"Standard deviation of the salary (sigma): {std_salary}")


Standard deviation of the salary (sigma): 16447.25


## Summary

- After removing Vologda Region and Murmansk Region, the sample mean salary (~38,658)
  is noticeably higher than the median (~32,473). This gap between mean and median is a
  classic sign of a **right-skewed distribution** — a small number of regions (e.g.
  Moscow, oil/gas regions) have unusually high average salaries and pull the mean
  upward, while most regions cluster at lower values.
- The large variance / standard deviation (~16,447) confirms there is substantial
  spread in regional salaries across Russia.
- Takeaway for later exercises: whenever mean and median disagree by a lot, it's worth
  plotting the distribution (e.g. a histogram) before trusting the mean as a "typical"
  value.
